hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,551 @@
+            final var senderId = nftTransfer.senderAccountIDOrElse(AccountID.DEFAULT);
+            validateAccountID(senderId, null);
+            checkSender(senderId, nftTransfer, context, accountStore);
+            checkPayer(senderId, context);
+            final var senderAccount = accountStore.getAliasedAccountById(senderId);
+            final var tokenRel = tokenRelStore.get(senderId, tokenID);
+            validateTruePreCheck(tokenRel != null, TOKEN_NOT_ASSOCIATED_TO_ACCOUNT);
+
+            final var receiverId = nftTransfer.receiverAccountIDOrElse(AccountID.DEFAULT);
+            validateAccountID(receiverId, null);
+            checkReceiver(receiverId, senderId, nftTransfer, context, tokenMeta, null, accountStore);
+            final var receiverAccount = accountStore.getAliasedAccountById(receiverId);
+
+            if (senderAccount == null || receiverAccount == null) {
+                throw new PreCheckException(INVALID_TRANSACTION_BODY);
+            }
+
+            final var nft = nftStore.get(tokenID, nftTransfer.serialNumber());
+            validateTrue(nft != null, INVALID_NFT_ID);
+
+            if (nftTransfer.isApproval()) {
+                // If isApproval flag is set then the spender account must have paid for the transaction.
+                // The transfer list specifies the owner who granted allowance as sender
+                // check if the allowances from the sender account has the payer account as spender
+                validateSpenderHasAllowance(senderAccount, context.payer(), tokenID, nft);
+            }
+
+            // owner of nft should match the sender in transfer list
+            if (nft.hasOwnerId()) {
+                validateTrue(nft.ownerId() != null, INVALID_NFT_ID);
+                validateTrue(nft.ownerId().equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
+            } else {
+                final var treasuryId = token.treasuryAccountId();
+                validateTrue(treasuryId != null, INVALID_ACCOUNT_ID);
+                validateTrue(treasuryId.equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
+            }
+        }
+    }
+
+    private void validateSpenderHasAllowance(
+            final Account owner, final AccountID spender, final TokenID tokenId, final Nft nft) {
+        final var approveForAllAllowances = owner.approveForAllNftAllowances();
+        final var allowance = AccountApprovalForAllAllowance.newBuilder()
+                .spenderId(spender)
+                .tokenId(tokenId)
+                .build();
+        if (!approveForAllAllowances.contains(allowance)) {
+            final var approvedSpender = nft.spenderId();
+            validateTrue(approvedSpender != null && approvedSpender.equals(spender), SPENDER_DOES_NOT_HAVE_ALLOWANCE);
+        }
+    }
+
+    private boolean tokenHasNoCustomFeesPaidByReceiver(TokenID tokenId, ReadableTokenStore tokenStore) {
+        final var token = getIfUsable(tokenId, tokenStore, PERMIT_PAUSED);
```

Comment:
Do we allow paused tokens to be airdropped?

Reply:
Makes sense that we don't, copy paste error, will fix it.